lib/core/test/unit/Cardano/Wallet/Primitive/Types/UTxOSpec.hs

```diff
@@ -78,6 +93,27 @@ spec =
         it "filterByAddress is always subset" $
             property prop_filterByAddress_isSubset
 
+prop_deltaUTxO_semigroup_apply :: Property
+prop_deltaUTxO_semigroup_apply = 
+        delta2 `apply` (delta1 `apply` utxo0)
+    ===
+        (delta2 <> delta1) `apply` utxo0
+  where
+    -- As the functions in 'DeltaUTxO' are implemented in terms of set
+    -- operations, we only have to test a Venn diagram using few elements.
+    utxo0  = mkUTxOs ["a0","a1","a2"]
```

Comment:
I like this manual example as it shows the case. One case. But what about also using proper arbitrary instances showing that deltaUTxO ops works properly for different overlapping of deltas? There are many possible cases here, right?

Reply:
The beauty of elementary statements about sets is that they can be decided by looking at truth tables, which in turn corresponds to one universal example ("Venn diagram").

For example, in order to show that the equality `(A ∩ B) ∪ C = (A ∪ C) ∩ (B ∪ C)` holds for all finite sets A,B,C, it is sufficient to show that it holds for the specific choice `A = { "100", "110", "101", "111"}`, `B = { "010", "110", "011", "111"}`, and `C = { "001", "101", "011", "111"}`. Even though the elements like "010" seem very specific, they stand for an entire set of elements; here, "010" stands for all elements that are contained in `B`, but not in `A` or `C`.

The associativity of `DeltaUTxO` is a statement of this form, assuming that all operations are implemented in terms of set-theoretic operations like `Set.intersect`, which they essentially are. Hence, we can decide its validity with a single, universal example — the one in the test. This example can be visualize as follows:

![venn diagram](https://user-images.githubusercontent.com/564592/156555696-06984aae-644f-4566-9d8b-f12be12c9d4b.png)

In this sense, "a2" represents an entire set of outputs (suggested by the colored dots), namely the set of outputs contained in `utxo0`, not spent by `delta1`, but spent by `delta2`. Likewise, "a0" represents the set of outputs contained `utxo0`, not spent by `delta1`, and not spent by `delta2`. Due to the "no double-spending" constraint, some sets must be empty. The visual diagrams shows all possible cases.